src/main/java/todoify/storage/InternalPath.java

```diff
@@ -0,0 +1,101 @@
+package todoify.storage;
+
+import java.io.File;
+import java.nio.file.Path;
+import java.util.Arrays;
+
+/**
+ * An internal path is a class that represents the path from an arbitrary base directory.
+ */
+public class InternalPath {
+
+    private static final Path DEFAULT_BASE_JAVA_NIO_FILE_PATH = Path.of(".", "todoifydata");
+
+    private final String[] components;
+
+    /**
+     * Constructs a new path with path components representing the internal relative path starting from the base
+     * directory.
+     *
+     * @param components The path components.
+     */
+    private InternalPath(String... components) {
+        this.components = components;
+    }
+
+    /**
+     * Constructs a new path with path components representing the internal relative path starting from the base
+     * directory.
+     *
+     * <p>
+     * All components are separated by path separators automatically, do not supply them. Instead, supply each component
+     * in the path as a separate argument.
+     * </p>
+     *
+     * @param components The path components.
+     */
+    public static InternalPath of(String... components) {
+        return new InternalPath(components);
+    }
+
+    /**
+     * Returns the internal relative path components, representing the path if joined with a path separator.
+     *
+     * @return The internal relative path components.
+     */
+    public String[] getComponents() {
+        return Arrays.copyOf(this.components, this.components.length);
+    }
+
+    /**
+     * Returns a new path relative to the configured base directory, while removing the last component in the path.
+     *
+     * @return The path after removing the last path component.
+     * @throws NegativeArraySizeException if there are no more components in the path to strip.
+     */
+    public InternalPath excludingLastComponent() {
+        String[] pathComponents = Arrays.copyOf(this.components, this.components.length - 1);
+        return new InternalPath(pathComponents);
+    }
+
+    /**
+     * Returns a new path relative to the configured base directory, while adding the given last component to the path
+     *
+     * @param component The new component to append to the end of the path.
+     * @return The path after adding the new last path component.
+     */
+    public InternalPath appendingLastComponent(String component) {
```

Comment:
Thanks for highlighting this. I think I would rename this to `getWithAppendedLastComponent` instead, and similarly for other methods, to comply with our conventions.

---

The reason this was named in continuous tense was actually based on a different naming convention to imply that the "action" is not performed on the instance in question, but rather returns a new instance, such as the following:
- [Apple's Swift: `URL#appendPathComponent`](https://developer.apple.com/documentation/foundation/url/1780397-appendpathcomponent), which updates the instance in place
- [Apple's Swift: `URL#appendingPathComponent`](https://developer.apple.com/documentation/foundation/nsurl/1410614-appendingpathcomponent), which returns a new instance

This would reduce confusion as to whether an action is self-mutating or returns a new copy, since present tense often implies the former in a number of languages.